client-common/src/tendermint/websocket_rpc_client.rs

```diff
@@ -0,0 +1,366 @@
+use serde::Deserialize;
+use serde_json::{json, Value};
+use websocket::sender::Writer;
+use websocket::stream::sync::TcpStream;
+use websocket::OwnedMessage;
+
+use self::types::*;
+use crate::tendermint::types::*;
+use crate::tendermint::Client;
+use crate::{Error, ErrorKind, Result, ResultExt};
+
+const RESPONSE_TIMEOUT: Duration = Duration::from_secs(10);
+
+const WAIT_FOR_CONNECTION_SLEEP_INTERVAL: Duration = Duration::from_millis(200);
+const WAIT_FOR_CONNECTION_COUNT: usize = 50;
+
+/// Tendermint RPC Client (uses websocket in transport layer)
+#[derive(Clone)]
+pub struct WebsocketRpcClient {
+    connection_state: Arc<Mutex<ConnectionState>>,
+    websocket_writer: Arc<Mutex<Writer<TcpStream>>>,
+    channel_map: Arc<Mutex<HashMap<String, SyncSender<JsonRpcResponse>>>>,
+}
+
+impl WebsocketRpcClient {
+    /// Creates a new instance of `WebsocketRpcClient`
+    //
+    // # How it works
+    //
+    // - Spawns `websocket_rpc_loop`.
+    // - Spawns `websocket_rpc_loop` monitor.
+    pub fn new(url: &str) -> Result<Self> {
+        let channel_map: Arc<Mutex<HashMap<String, SyncSender<JsonRpcResponse>>>> =
+            Default::default();
+
+        let (websocket_reader, websocket_writer) = websocket_rpc_loop::new_connection(url)?;
+        let websocket_writer = Arc::new(Mutex::new(websocket_writer));
+
+        let loop_handle = websocket_rpc_loop::spawn(
+            channel_map.clone(),
+            websocket_reader,
+            websocket_writer.clone(),
+        );
+
+        let connection_state = websocket_rpc_loop::monitor(
+            url.to_owned(),
+            channel_map.clone(),
+            loop_handle,
+            websocket_writer.clone(),
+        );
+
+        Ok(Self {
+            connection_state,
+            websocket_writer,
+            channel_map,
+        })
+    }
+
+    /// Returns current connection state of websocket connection
+    pub fn connection_state(&self) -> ConnectionState {
+        *self
+            .connection_state
+            .lock()
+            .expect("Unable to acquire lock on connection state")
+    }
+
+    /// Sends a RPC request
+    //
+    // # How it works
+    //
+    // - Prepares JSON-RPC request from `method` and `params` (generates a random `request_id`).
+    // - Creates a `sync_channel` pair
+    // - Inserts `channel_sender` to `channel_map` corresponding to generated `request_id`.
+    // - Ensure that the `websocket_rpc_loop` is in `Connected` state.
+    // - Send request websocket message.
+    // - Receive response on `channel_receiver`.
+    fn request(&self, method: &str, params: &[Value]) -> Result<Value> {
+        let (id, channel_receiver) = self.send_request(method, params)?;
+        self.receive_response(method, params, &id, channel_receiver)
+    }
+
+    /// Sends RPC requests for a batch.
+    ///
+    /// # Note
+    ///
+    /// This does not use batch JSON-RPC requests but makes multiple single JSON-RPC requests in parallel.
```

Comment:
why not to use the batch request?

Reply:
what happens? does it crash?
I think the JSON-RPC handler is the same for both transports

Reply:
I returns error saying that it is unable to parse request json. Anyway, it does not matter on websockets because the amount of data transferred in batch request vs multiple requests in exactly same (as there is no overhead of doing a handshake for each request).